src/ucp/wireup/wireup_cm.c

```diff
@@ -229,37 +227,226 @@ static void ucp_cm_client_connect_cb(uct_ep_h ep, void *arg,
 
     if (addr.address_count == 0) {
         status = UCS_ERR_UNREACHABLE;
-        goto free_addr_list;
+        goto out_free_addr;
     }
 
-    ucs_assert(addr.address_count <= UCP_MAX_RESOURCES);
     for (addr_idx = 0; addr_idx < addr.address_count; ++addr_idx) {
-        addr.address_list[addr_idx].dev_addr = remote_data->dev_addr;
+        addr.address_list[addr_idx].dev_addr = progress_arg->dev_addr;
     }
 
+    UCS_ASYNC_BLOCK(&worker->async);
+
+    ucp_ep_update_dest_ep_ptr(ucp_ep, progress_arg->sa_data->ep_ptr);
+
+    ucs_assert(addr.address_count <= UCP_MAX_RESOURCES);
     ucs_assert(wireup_ep->ep_init_flags & UCP_EP_INIT_CM_WIREUP_CLIENT);
     status = ucp_wireup_init_lanes(ucp_ep, wireup_ep->ep_init_flags, &addr,
                                    addr_indices);
     if (status != UCS_OK) {
-        goto free_addr_list;
+        goto out_unblock;
+    }
+
+    status = ucp_wireup_connect_local(ucp_ep, &addr, NULL);
+    if (status != UCS_OK) {
+        goto out_unblock;
     }
 
-    ucp_wireup_connect_local(ucp_ep, &addr, NULL);
     ucp_wireup_remote_connected(ucp_ep);
 
-free_addr_list:
+out_unblock:
+    UCS_ASYNC_UNBLOCK(&worker->async);
+out_free_addr:
     ucs_free(addr.address_list);
 out:
+    ucs_free(progress_arg->sa_data);
+    ucs_free(progress_arg->dev_addr);
+    ucs_free(progress_arg);
+
     if (status != UCS_OK) {
-        ucp_worker_set_ep_failed(worker, ucp_ep,
-                                 wireup_ep ? &wireup_ep->super.super : ep,
+        ucp_worker_set_ep_failed(worker, ucp_ep, &wireup_ep->super.super,
                                  ucp_ep_get_cm_lane(ucp_ep), status);
     }
+
+    return 1;
+}
+
+static void ucp_cm_client_connect_cb(uct_ep_h uct_cm_ep, void *arg,
+                                     const uct_cm_remote_data_t *remote_data,
+                                     ucs_status_t status)
+{
+    ucp_ep_h ucp_ep            = (ucp_ep_h)arg;
+    ucp_worker_h worker        = ucp_ep->worker;
+    uct_worker_cb_id_t prog_id = UCS_CALLBACKQ_ID_NULL;
+    ucp_cm_client_connect_progress_arg_t *progress_arg;
+
+    if (status != UCS_OK) {
```

Comment:
should the ep be marked here as failed somehow?

Reply:
yes, see `err_out:` label in the end of the function